File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/model/SystemItemParser.java
package org.fbme.smvDebugger.model;

import java.util.Objects;
import org.fbme.smvDebugger.commons.ArrayUtils;

public class SystemItemParser {
  private static final String FULL_NAME_DELIMITER = "[.]";
  private static final String EVENT_PORT_DELIMITER = "_";
  private static final String TIME_EVENT_PORT_VALUE = "value";
  private static final String EVENT_PORT_KEY_FEATURE = "event";
  private static final String ECC_KEY_FEATURE = "Q_smv";
  private static final String ECC_ITEM_NAME = "ECC";

  public SystemItem parse(final String fullName) {
    final String[] nameParts = fullName.split(FULL_NAME_DELIMITER);

    final SystemItemType type;
    if (isEventPort(nameParts)) {
      type = SystemItemType.EVENT_PORT;
    } else if (isDataPort(nameParts)) {
      type = SystemItemType.DATA_PORT;
    } else if (isECC(nameParts)) {
      type = SystemItemType.ECC;
    } else {
      return null;
    }

    final String[] fbNames = getFbNames(nameParts);
    if (fbNames.length == 0) {
      return null;
    }

    final String itemName = getItemName(nameParts, type);
    return new SystemItem(fullName, fbNames, itemName, type);
  }

  private boolean isEventPort(final String[] nameParts) {
    return isPort(nameParts) && nameParts[nameParts.length - 1].contains(EVENT_PORT_KEY_FEATURE);
  }

  private boolean isDataPort(final String[] nameParts) {
    return isPort(nameParts) && !((isEventPort(nameParts)));
  }

  private boolean isECC(final String[] nameParts) {
    return Objects.equals(nameParts[nameParts.length - 1], ECC_KEY_FEATURE);
  }

  private boolean isPort(final String[] nameParts) {
    return nameParts.length > 1 && !(isECC(nameParts));
  }

  private String[] getFbNames(final String[] nameParts) {
    return ArrayUtils.subarray(nameParts, 1, nameParts.length - 1);
  }

  private String getItemName(final String[] nameParts, final SystemItemType type) {
    switch (type) {
      case EVENT_PORT:
        return nameParts[nameParts.length - 1].split(EVENT_PORT_DELIMITER)[1];
      case DATA_PORT:
        return nameParts[nameParts.length - 1];
      case ECC:
        return ECC_ITEM_NAME;
      default:
        return null;
    }
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/model/SystemItemParser.kt
package org.fbme.smvDebugger.model

import org.fbme.smvDebugger.commons.ArrayUtils

class SystemItemParser {
    fun parse(fullName: String): SystemItem? {
        val nameParts = fullName.split(Regex(FULL_NAME_DELIMITER)).toTypedArray()
        val type: SystemItemType = if (isEventPort(nameParts)) {
            SystemItemType.EVENT_PORT
        } else if (isDataPort(nameParts)) {
            SystemItemType.DATA_PORT
        } else if (isECC(nameParts)) {
            SystemItemType.ECC
        } else {
            return null
        }
        val fbNames = getFbNames(nameParts)
        if (fbNames.isEmpty()) {
            return null
        }
        val itemName = getItemName(nameParts, type)
        return SystemItem(fullName, fbNames, itemName, type)
    }

    private fun isEventPort(nameParts: Array<String>): Boolean {
        return isPort(nameParts) && nameParts[nameParts.size - 1].contains(EVENT_PORT_KEY_FEATURE)
    }

    private fun isDataPort(nameParts: Array<String>): Boolean {
        return isPort(nameParts) && !isEventPort(nameParts)
    }

    private fun isECC(nameParts: Array<String>): Boolean {
        return nameParts[nameParts.size - 1] == ECC_KEY_FEATURE
    }

    private fun isPort(nameParts: Array<String>): Boolean {
        return nameParts.size > 1 && !isECC(nameParts)
    }

    private fun getFbNames(nameParts: Array<String>): Array<String> {
        return ArrayUtils.subarray(nameParts, 1, nameParts.size - 1)
    }

    private fun getItemName(nameParts: Array<String>, type: SystemItemType): String {
        return when (type) {
            SystemItemType.EVENT_PORT -> nameParts[nameParts.size - 1]
                .split(EVENT_PORT_DELIMITER)
                .toTypedArray()[1]

            SystemItemType.DATA_PORT -> nameParts[nameParts.size - 1]
            SystemItemType.ECC -> ECC_ITEM_NAME
        }
    }

    companion object {
        private const val FULL_NAME_DELIMITER = "[.]"
        private const val EVENT_PORT_DELIMITER = "_"
        private const val TIME_EVENT_PORT_VALUE = "value"
        private const val EVENT_PORT_KEY_FEATURE = "event"
        private const val ECC_KEY_FEATURE = "Q_smv"
        private const val ECC_ITEM_NAME = "ECC"
    }
}
